janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java

```diff
@@ -546,13 +572,42 @@ private Response performRequest(String method, String path, byte[] requestData)
         return performRequest(new Request(method, path), requestData);
     }
 
+    private Response performRequestWithRetry(Request request) throws IOException {
+        int retryCount = 0;
+        ResponseException lastException;
+        do {
+            if (retryCount > 0) {
+                //Wait before trying again
+                long waitDurationMs = Math.min((long) (retryInitialWaitMs * Math.pow(10, retryCount - 1)), retryMaxWaitMs);
+                log.warn("Retrying Elasticsearch request in {} ms. Attempt {} of {}", waitDurationMs, retryCount, retryAttemptLimit);
+                try {
+                    Thread.sleep(waitDurationMs);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(String.format("Thread interrupted while waiting for retry attempt %d of %d", retryCount, retryAttemptLimit), e);
+                }
+            }
+            try {
+                return delegate.performRequest(request);
+            } catch (ResponseException e) {
+                if (!retryOnErrorCodes.contains(e.getResponse().getStatusLine().getStatusCode())) {
+                    throw e;
+                }
+                lastException = e;
+            }
+            retryCount++;
+        } while (retryCount <= retryAttemptLimit);
+
+        //If we've exceeded our retry limit then throw the last exception that was captured from retrying
+        throw lastException;
```

Comment:
I don't follow how it's a bug @li-boxuan.

> If the delegate.performRequest(request) does not throw a ResponseException on the first attempt and there are no retry attempts (because the response is successful or an error not suitable for retry occurs), the do-while loop will not enter or will exit early, and lastException will never be assigned a value.

ChatGPT seems to have missed the semantic implication that other exception types are not caught and thus would be thrown _immediately_ out of the method. In that scenario `lastException` never being assigned a value is moot.

The four cases should be:

1. The request is successful - we return out of the method.
2. The request is not successful - but it's an exception that's different than a `ResponseException` so it immediately gets thrown up out of the method.
3. The request is not successful - it is a `ResponseException` _but_ it is not a configured error code for retry and thus will be immediately thrown up out of the method.
4. The request is not successful - it is a `ResponseException` _and_ it *is* a configured error code for retry, which will then cause `lastException` to be assigned and possibly later thrown if it is the last (or only) attempt if retries are set to 0.

Reply:
Added some more tests in 9b57c0349 to verify what ChatGPT is claiming is a non-issue.

Reply:
Oh lol thank you so much for the detailed explanation. I rely on AI too much these days... and got cheated by its hallucination that at the first glance, seemed reasonable.